goasupport/generator.go

```diff
@@ -0,0 +1,292 @@
+			return nil, err
+		}
+		if err := ctxWr.ExecuteTemplate("generateETag", generateETag, nil, entity); err != nil {
+			return nil, err
+		}
+	}
+	err = ctxWr.FormatCode()
+	if err != nil {
+		return nil, err
+	}
+	return []string{ctxFile}, nil
+}
+
+const (
+	headerMethods = `
+{{ range $req := . }}{{ range $head := $req.Headers }}
+// Get{{ $head.Name }} return the HTTP Header {{ $head.Header }}
+func (ctx *{{$req.Name}}) Get{{ $head.Name }}() *{{ $head.Type }} {
+	return ctx.{{ $head.Name }}
+}{{ end }}
+{{ end }}
+`
+	genericOKMethod = `
+{{ $resp := . }}
+// GenericOK returns the '200 OK' response with the given entity in the body
+// and casts it to '{{$resp.Entity.TypeName}}' before calling the '{{$resp.Name}}.OK({{$resp.Entity.TypeName}})' method
+func (ctx *{{$resp.Name}}) GenericOK(entity interface{}) error {
+	res := entity.({{$resp.Entity.TypeName}})
+	return ctx.OK(&res)
+}`
+
+	setHeader = `
+{{ $resp := . }}
+// SetHeader sets the header with the given name and value
+func (ctx *{{$resp.Name}}) SetHeader(name, value string) {
+	ctx.ResponseData.Header().Set(name, value)
+}`
+
+	modifiedSince = `
+{{ $resp := . }}
+// ModifiedSince returns 'true' if the given 'lastModified' argument is after the context's 'IfModifiedSince' value
+func (ctx *{{$resp.Name}}) ModifiedSince(lastModified time.Time) bool {
+	if ctx.IfModifiedSince != nil {
+		ifModifiedSince := ctx.IfModifiedSince.UTC()
+		return ifModifiedSince.Before(lastModified.UTC())
+	}
+	return true
+}`
+
+	matchesETag = `
+{{ $resp := . }}
+// MatchesETag returns 'true' the given 'etag' argument matches with the context's 'IfNoneMatch' value.
+func (ctx *{{$resp.Name}}) MatchesETag(etag string) bool {
+	if ctx.IfNoneMatch != nil && *ctx.IfNoneMatch == etag {
+		return true
+	}
+	return false
+}`
+
+	getLastModified = `
+{{ $entity := . }}
+{{ if $entity.IsSingle }}
+ // GetLastModified gets the update time for a given element.
+func (entity {{$entity.TypeName}}) GetLastModified() time.Time {
+	var updatedAt time.Time
+	if entity.Data.Attributes.UpdatedAt != nil && entity.Data.Attributes.UpdatedAt.After(updatedAt) {
+		updatedAt = *entity.Data.Attributes.UpdatedAt
+	}
+	return updatedAt.Truncate(time.Second).UTC()
+}
+{{ end }}
+{{ if $entity.IsList }}
+// GetLastModified gets the update time for a given element.
+func (entity {{$entity.TypeName}}) GetLastModified() time.Time {
+	var updatedAt time.Time
+	for _, data := range entity.Data {
+		if data.Attributes.UpdatedAt != nil && data.Attributes.UpdatedAt.After(updatedAt) {
+			updatedAt = *data.Attributes.UpdatedAt
+		}
+	}
+	return updatedAt.Truncate(time.Second).UTC()
+}
+{{ end }}`
+
+	generateETag = `
+{{ $entity := . }}
+{{ if $entity.IsSingle }}
+// GenerateETag generates the value to return in the "ETag" HTTP response header, using the data in the given buffer.
+// The ETag is the base64-encoded value of the md5 hash of the buffer content
+func (entity {{$entity.TypeName}}) GenerateETag() string {
```

Comment:
GenerateETag is too late to put it on the app.X classes used in the API. We've already done all the work at that point. We need them based on the 'internal db models' or at least as a layer between. Then we can make the conditional() conditionally trigger the conversion.

We could make it an interface the 'Entities' cohere to; 

```
func (s User) Tag() []interface{}{
   return []interface{s.ID, s.UpdatedAt}
}